src/compas/datastructures/assembly/part.py

```diff
@@ -0,0 +1,187 @@
+        The base shape of the part geometry.
+    features : list of tuple(:class:`compas.geometry.Shape`, str), optional
+        The features to be added to the base shape of the part geometry.
+
+    Attributes
+    ----------
+    attributes : dict
+        General object attributes that will be included in the data dict.
+    key : int or str
+        The identifier of the part in the connectivity graph of the parent assembly.
+    frame : :class:`compas.geometry.Frame`
+        The local coordinate system of the part.
+    shape : :class:`compas.geometry.Shape`
+        The base shape of the part geometry.
+    features : list of tuple(:class:`compas.geometry.Shape`, str)
+        The features added to the base shape of the part geometry.
+    transformations : deque of :class:`compas.geometry.Transformation`
+        The stack of transformations applied to the part geometry.
+        The most recent transformation is on the left of the stack.
+        All transformations are with respect to the local coordinate system.
+    """
+
+    operations = {
+        'union': boolean_union_mesh_mesh,
+        'difference': boolean_difference_mesh_mesh,
+        'intersection': boolean_intersection_mesh_mesh
+    }
+
+    @property
+    def DATASCHEMA(self):
+        import schema
+        return schema.Schema({
+            "attributes": dict,
+            "key": int,
+            "frame": Frame,
+            "shape": Shape,
+            "features": list,
+            "transformations": list,
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        return 'part'
+
+    def __init__(self, name, frame=None, shape=None, features=None, **kwargs):
+        super(Part, self).__init__()
+        self._frame = None
+        self.attributes = {'name': name or 'Part'}
+        self.key = None
+        self.frame = frame
+        self.shape = shape or Shape([], [])
+        self.features = features or []
+        self.transformations = deque()
+
+    def __str__(self):
+        tpl = "<Part with shape {} and features {}>"
+        return tpl.format(self.shape, self.features)
+
+    @property
+    def data(self):
+        """dict : A data dict representing the part attributes, the assembly graph identifier, the local coordinate system,
+        the base shape, the shape features, and the transformation tack wrt to the local coordinate system.
+        """
+        data = {
+            'attributes': self.attributes,
+            "key": self.key,
+            "frame": self.frame.data,
+            "shape": self.shape.data,
+            "features": [(shape.data, operation) for shape, operation in self.features],
+            "transformations": [T.data for T in self.transformations],
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self.attributes.update(data['attributes'] or {})
+        self.key = data['key']
+        self.frame.data = data['frame']
+        self.shape.data = data['shape']
```

Comment:
this doesn't deserialize properly in my tests... If I change it to rely on the recursive serialization thou, it works:

```suggestion
            "frame": self.frame,
            "shape": self.shape,
            "features": [(shape.data, operation) for shape, operation in self.features],
            "transformations": [T.data for T in self.transformations],
        }
        return data

    @data.setter
    def data(self, data):
        self.attributes.update(data['attributes'] or {})
        self.key = data['key']
        self.frame = data['frame']
        self.shape= data['shape']
```

Reply:
the explicit data conversion is necessary through. so something else must be wrong. i assume you used the json serializers of the data framework. this will indeed work, but the data dict should be a pure Python representation of the object even without serialisation... without the use of `.data` it is basically a collection of COMPAS objects.